grid-client/deployer/deployment_deployer.go

```diff
@@ -185,24 +212,204 @@ func (d *DeploymentDeployer) updateStateFromDeployments(ctx context.Context, dl
 		}
 
 		if dlData.Name == dl.Name {
-			dl.NodeDeploymentID[dl.NodeID] = newDl.ContractID
+			if newDl.ContractID != 0 {
+				dl.NodeDeploymentID[dl.NodeID] = newDl.ContractID
+				dl.ContractID = newDl.ContractID
+				d.tfPluginClient.State.StoreContractIDs(dl.NodeID, []uint64{dl.ContractID})
+			}
 		}
 	}
 
-	if contractID, ok := dl.NodeDeploymentID[dl.NodeID]; ok && contractID != 0 {
-		dl.ContractID = contractID
-		if !workloads.Contains(d.tfPluginClient.State.CurrentNodeDeployments[dl.NodeID], dl.ContractID) {
-			d.tfPluginClient.State.CurrentNodeDeployments[dl.NodeID] = append(d.tfPluginClient.State.CurrentNodeDeployments[dl.NodeID], dl.ContractID)
+	return nil
+}
+
+func (d *DeploymentDeployer) getUsedHostIDsOfNodeWithinNetwork(ctx context.Context, nodeID uint32, networkName string) ([]byte, error) {
+	var usedHostIDs []byte
+
+	nodeClient, err := d.tfPluginClient.NcPool.GetNodeClient(d.tfPluginClient.SubstrateConn, nodeID)
+	if err != nil {
+		return nil, errors.Wrapf(err, "could not get node client for node %d", nodeID)
+	}
+
+	privateIPs, err := nodeClient.NetworkListPrivateIPs(ctx, networkName)
+	if err != nil {
+		return nil, errors.Wrapf(err, "could not list private ips from node %d", nodeID)
+	}
+
+	network := d.tfPluginClient.State.Networks.GetNetwork(networkName)
+	ipRange := network.GetNodeSubnet(nodeID)
+
+	_, ipRangeCIDR, err := net.ParseCIDR(ipRange)
+	if err != nil {
+		return nil, errors.Wrapf(err, "invalid ip range %s", ipRange)
+	}
+
+	// get used host IDs of the deployment node using the private IPs list of the node
+	for _, privateIP := range privateIPs {
+		parsedPrivateIP := net.ParseIP(privateIP).To4()
+
+		if parsedPrivateIP == nil {
+			log.Debug().Err(fmt.Errorf("[Error]: network %s has a nil private ip %s", networkName, privateIP)).Send()
```

Comment:
it's a nitpick but it's not necessarily a nil private ip but could be any random characters that are not valid ipv4 :D